src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,178 @@
+section \<open>Chain Synchronization Mini-Protocol\<close>
+
+text \<open>
+  The theory in this section implements the chain synchronization mini-protocol.
+\<close>
+
+theory "Ouroboros-Mini_Protocols-Chain_Sync"
+  imports
+    "Ouroboros-Mini_Protocols"
+    "HOL-Library.BNF_Corec"
+begin
+
+locale chain_sync =
+  fixes initial_items\<^sub>c :: "'i list"
+  fixes initial_items\<^sub>s :: "'i list"
+  fixes candidate_points :: "'i list \<Rightarrow> 'p list"
+  fixes best_intersection_point :: "'i list \<Rightarrow> 'p list \<Rightarrow> 'p option"
+  fixes point_of :: "'i \<Rightarrow> 'p"
+
+subsection \<open>Parties\<close>
+
+datatype party =
+  Client |
+  Server
+
+subsection \<open>State Machine\<close>
+
+datatype state =
+  Idle |
+  Intersect |
+  CanAwait
+
+datatype ('i, 'p) message =
+  is_next_request: RequestNext |
+  is_roll_forward: RollForward \<open>'i\<close> |
+  is_roll_backward: RollBackward \<open>'p\<close> |
+  is_await_reply: AwaitReply |
+  is_find_intersect: FindIntersect \<open>'p list\<close> |
+  is_intersect_found: IntersectFound \<open>'p\<close> |
+  is_intersect_not_found: IntersectNotFound
+
+fun agent_in_state' where
+  "agent_in_state' Idle = Client" |
+  "agent_in_state' _ = Server"
+
+inductive can_finish_in_state' where
+  "can_finish_in_state' Idle"
+
+declare can_finish_in_state'.simps [simp]
+
+primrec next_state' where
+  "next_state' Idle m = (partial_case m of
+    FindIntersect _ \<Rightarrow> Intersect |
+    RequestNext \<Rightarrow> CanAwait
+  )" |
+  "next_state' Intersect m = (partial_case m of
+    IntersectFound _ \<Rightarrow> Idle |
+    IntersectNotFound \<Rightarrow> Idle
+  )" |
+  "next_state' CanAwait m = (partial_case m of
+    RollForward _ \<Rightarrow> Idle |
+    RollBackward _ \<Rightarrow> Idle |
+    AwaitReply \<Rightarrow> Idle \<comment> \<open>only for this initial implementation\<close>
+  )"
+
+definition state_machine where
+  [simp]: "state_machine = \<lparr>
+    initial_state = Idle,
+    agent_in_state = agent_in_state',
+    can_finish_in_state = can_finish_in_state',
+    next_state = next_state'
+  \<rparr>"
+
+sublocale chain_sync \<subseteq> protocol_state_machine \<open>state_machine\<close> .
+
+subsection \<open>Programs\<close>
+
+datatype phase =
+  is_intersection_finding: IntersectionFinding |
+  is_items_update: ItemsUpdate
+
+definition roll_back_to :: "'i list \<Rightarrow> 'p \<Rightarrow> ('i \<Rightarrow> 'p) \<Rightarrow> 'i list" where
+  [simp]: "roll_back_to \<I> p \<psi> = (THE \<I>\<^sub>1. \<exists>\<I>\<^sub>2. \<I> = \<I>\<^sub>1 @ \<I>\<^sub>2 \<and> \<psi> (last \<I>\<^sub>1) = p)"
```

Comment:
>I think the argument ψ :: 'i ⇒ 'p should come first.

Fixed in 0b5be84cea502543302d162673330fa57af1604f, please check.

Reply:
>Therefore, I suggest to drop the preposition and thus go for simply roll_back.
Fixed in 835363b0bbdeb4dab1e5af1fa5b6e8299c74c843, please check.

Reply:
>I think that here and in the rest of this theory we should use the identifier I instead of ℐ

As part of a687abbb107eb649e3560199bf4c652024f5537c, I changed ℐ to 𝒞. Is that fine with you?

Reply:
>We should put parentheses around ∃ℐ⇩2. ℐ = ℐ⇩1 @ ℐ⇩2. 

Fixed in ef9c9612a8ac60a2793c4090f820d275af46be18, please check:

Note: The function `prefix` from `Sublist` is the Isabelle/HOL equivalent of Haskell's `isPrefixOf`.

Reply:
> > I think that here and in the rest of this theory we should use the identifier `I` instead of `ℐ` […].
> 
> As part of [a687abb](https://github.com/input-output-hk/ouroboros-high-assurance/commit/a687abbb107eb649e3560199bf4c652024f5537c), I changed `ℐ` to `𝒞`. Is that fine with you?

Well, at first sight, chains are just lists of items, which means that `I` would be the natural choice as an identifier. However, a chain is actually more than a mere list of items, as the items have to be related in certain ways in order for the list of items representing a correct chain (and in other contexts there is also the separate genesis item, which is different from ordinary items). Therefore, using the letter `c` seems appropriate. It might even be a good idea to use its uppercase version, `C`, in order to signify that a chain is a more complex data structure (for example, I use `P` for possibilities, although in this case the name of the data structure, *possibilities*, is a plural). However, why additionally use calligraphic style? Is `𝒞` used for chains in Ouroboros papers?

Reply:
> > We should put parentheses around `∃ℐ⇩2. ℐ = ℐ⇩1 @ ℐ⇩2`.
> 
> Fixed in [ef9c961](https://github.com/input-output-hk/ouroboros-high-assurance/commit/ef9c9612a8ac60a2793c4090f820d275af46be18), please check.
> 
> Note: The function `prefix` from `Sublist` is the Isabelle/HOL equivalent of Haskell's `isPrefixOf`.

Who would have thought that there is a dedicated theory for sublists? Well, given that there is the `prefix` function, we should use it. Note that the right-hand side of its definition is exactly the explicit statement we’re using at the moment.

Reply:
> However, why additionally use calligraphic style? Is `𝒞` used for chains in Ouroboros papers?

Yes, that's the style used in Ouroboros papers. Would you suggest to keep this style or drop it in favor of the use of `C`?

Reply:
> Well, given that there is the `prefix` function, we should use it.

Fixed in 2da43a1a1d0f0c468a57a436d43174a1c52c9ecd, please check.

Reply:
> > Well, given that there is the `prefix` function, we should use it.
> 
> Fixed in [2da43a1](https://github.com/input-output-hk/ouroboros-high-assurance/commit/2da43a1a1d0f0c468a57a436d43174a1c52c9ecd), please check.

Looks good, except that, now that we don’t have a `𝒞⇩2` anymore, the name `𝒞⇩1` looks a bit out of place. Could you change it to `𝒞'` perhaps?

Reply:
> > However, why additionally use calligraphic style? Is `𝒞` used for chains in Ouroboros papers?
> 
> Yes, that's the style used in Ouroboros papers. Would you suggest to keep this style or drop it in favor of the use of `C`?

That’s a difficult question. It makes some sense to use the same notation as the papers, in particular where we write formal specifications of what the papers are describing, like your formal specification of Ouroboros Praos.

That said, I guess there will be more differences between the naming scheme the papers are using and the naming scheme we are using and actually have been using from the equivalence reasoner upwards. It’s simply not feasible to completely change our style to make it consistent with the style of the papers. Therefore, we should either stick to our style, come what may, or have some breaking points, where we switch from our notation to the notation of the papers.

I’m very much inclined to maintain a more or less consistent naming scheme and therefore opt for the former option. For this concrete situation here, let’s use `C` instead of `𝒞`.

Reply:
> > Well, given that there is the `prefix` function, we should use it.
> 
> Fixed in [2da43a1](https://github.com/input-output-hk/ouroboros-high-assurance/commit/2da43a1a1d0f0c468a57a436d43174a1c52c9ecd), please check.

Another thing regarding this fix: Maybe swap the order of the `Sublist` and the `BNF_Corec` import. After all, `BNF_Corec` is largely about what other systems would have as a language feature, while `Sublist` is clearly a library thing. Also, `BNF_Corec` is a theory that each of our mini-protocol specifications uses, which `Sublist` is not.

Reply:
> Looks good, except that, now that we don’t have a `𝒞⇩2` anymore, the name `𝒞⇩1` looks a bit out of place. Could you change it to `𝒞'` perhaps?

Ah, you're right, I forgot to change that. Fixed in 09127196e3deda071ed2d7d3f6aadbe4728b8688, please check.

Reply:
> Maybe swap the order of the `Sublist` and the `BNF_Corec` import

Fixed in 043166ee9d04cc272c50b4085ae8a5fc26f01539, please check.

Reply:
> For this concrete situation here, let’s use `C` instead of `𝒞`.

Fixed in 84464da97684e7afa77801070bb57004797d5f09, please check.